staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go

```diff
@@ -0,0 +1,418 @@
+type UserSubject struct {
+	// `name` is the username that matches, or "*" to match all usernames.
+	// Required.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+}
+
+// GroupSubject holds detailed information for group-kind subject.
+type GroupSubject struct {
+	// name is the user group that matches, or "*" to match all user groups.
+	// See https://github.com/kubernetes/apiserver/blob/master/pkg/authentication/user/user.go for some
+	// well-known group names.
+	// Required.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+}
+
+// ServiceAccountSubject holds detailed information for service-account-kind subject.
+type ServiceAccountSubject struct {
+	// `namespace` is the namespace of matching ServiceAccount objects.
+	// Required.
+	Namespace string `json:"namespace" protobuf:"bytes,1,opt,name=namespace"`
+	// `name` is the name of matching ServiceAccount objects, or "*" to match regardless of name.
+	// Required.
+	Name string `json:"name" protobuf:"bytes,2,opt,name=name"`
+}
+
+// ResourcePolicyRule is a predicate that matches some resource requests, testing the request's verb and the target
+// resource. A ResourcePolicyRule matches a request if and only if: (a) at least one member
+// of verbs matches the request, (b) at least one member of apiGroups matches the request, and (c) at least one member
+// of resources matches the request.
+type ResourcePolicyRule struct {
+	// `verbs` is a list of matching verbs and may not be empty.
+	// "*" represents all verbs. if it is present, it must be the only entry.
+	// Required.
+	Verbs []string `json:"verbs" protobuf:"bytes,1,rep,name=verbs"`
+	// `apiGroups` is a list of matching API groups and may not be empty.
+	// "*" represents all api-groups. if it is present, it must be the only entry.
+	// Required.
+	APIGroups []string `json:"apiGroups" protobuf:"bytes,2,rep,name=apiGroups"`
+	// `resources` is a list of matching resources (i.e., lowercase and plural) with, if desired, subresource.
+	// For example, [ "services", "nodes/status" ].
+	// "*" represents all resources. if it is present, it must be the only entry.
+	// Required.
+	Resources []string `json:"resources" protobuf:"bytes,3,rep,name=resources"`
+}
+
+// NonResourcePolicyRule is a predicate that matches non-resource requests according to their verb and the non-resource
+// URL they request to act upon. A NonResourcePolicyRule matches a request if and only if both (a) at least one member
+// of verbs matches the request and (b) at least one member of nonResourceURLs matches the request.
+type NonResourcePolicyRule struct {
+	// `verbs` is a list of matching verbs and may not be empty.
+	// "*" represents all verbs. if it is present, it must be the only entry.
+	// Required.
+	Verbs []string `json:"verbs" protobuf:"bytes,1,rep,name=verbs"`
+	// `nonResourceURLs` is a set of url prefixes that a user should have access to.
+	// For example, ["/apis", "/apis/*"].
```

Comment:
The examples so far raise the question of whether `"/api*"` would be allowed and have the obvious interpretation.

Reply:
Also we need to be clear about whether partial path elements are permitted, I'd go for "no"; you have to specify complete path elements, and * must be its own element, and the final one. E.g.
* "/ap*" is illegal
* "/ap" is legal but matches nothing
* "/ap/*" also matches nothing
* "/api/*" matches all core/v1 api paths

We should document what happens if you use a prefix which covers a resource, such as "/api/"; I expect resource accesses not to ever pass through here?

Reply:
Maybe /metrics, /healthz/*, etc would be better examples.

Reply:
updated the examples of partial urls